#Lecture 10 Live Coding Exercise

#The file needed for today's exercise is on OWl under Lab Resources ('EdgeR_RNA_all_genes.csv')

#Clear your environment
rm(list = ls())

#RNAseq example with a gene list
library(limma)
library(biomaRt)
library(dplyr)
library(AnnotationDbi)
library(enrichplot)
library(clusterProfiler)
library(DOSE)
library(annotables)
library(TimeSeriesExperiment)
library(org.Hs.eg.db)
library(pathview)

#We will start with the "summary statistics" from an experiment, or the results of a differential analysis. 

#First step when working with a microarray or similar is annotating your result to an annotation or manifest file to get your gene names

#setwd("/Users/Christina/Desktop/Western Assistant Professor/Teaching/MBI4650F")

#We will use a common example for RNA sequencing where your edgeR results have ENSG identifiers for genes:
#Seq <- read.csv("EdgeR_RNA_all_genes.csv", header=T) #edgeR results
Seq <- read.csv("/home/shared/Epigenomics/EdgeR_RNA_all_genes.csv", header=T)
names(Seq)[names(Seq) == "X"] <- "ensembl_gene_id" #rename first column

#Use Bonferroni correction for p-value
Ntests = nrow(Seq)
pval = 0.05/Ntests
pval
#3.441867e-06

#Filtering and joining with entrez IDs
Seq_sig <- merge(Seq, grch38, by.x="ensembl_gene_id", by.y="ensgene", sort=F)
Seq_sig <- filter(Seq_sig, PValue < pval)
Seq_sig <- Seq_sig[!is.na(Seq_sig$entrez),]
Seq_sig <- Seq_sig[order(Seq_sig$PValue),]
sig_genes <- Seq_sig$entrez

Seq_all <- merge(Seq, grch38, by.x="ensembl_gene_id", by.y="ensgene", sort=F)

length(sig_genes)
dim(Seq_all)

#The Importance of a 'universe'

#The background list or universe in a GO or KEGG analysis (or otherwise), is the list of all genes that you tested in your study. In the example of RNA sequencing, the universe would comprise all genes that were tested in your final model (in this case ~14000 genes)

#KEGG analysis
#The Kyoto Encyclopedia of Genes and Genomes (KEGG) database

#KEGG is a database resource for understanding high-level functions and utilities of the biological system, such as the cell, the organism and the ecosystem, from molecular-level information, especially large-scale molecular datasets generated by genome sequencing and other high-throughput experimental technologies.

#KEGG Enrichment
kegg_rna <- kegga(sig_genes, universe = Seq_all$entrez, species="Hs")
kegg_rna <- kegg_rna %>% arrange(P.DE)

kegg_rna$KEGG_label <- rownames(kegg_rna)
head(kegg_rna)

#Can pull top results with topKEGG
KEGG_enrich <- topKEGG(kegg_rna)
KEGG_enrich

#Extracting genes from a specific pathway
#Often times you would like to know what genes are in your significant gene list that are leading to an overenrichment result in your network/pathway analysis, here is one way to extract them.

#Extract genes for KEGG
GK <- getGeneKEGGLinks(species.KEGG = "hsa")
GKN <- subset(GK, GK$PathwayID == "path:hsa04080")
#362 genes in network, we tested 127 of them
GKN2 <- GKN$GeneID
#Your 23 genes in this pathway
SigNGenes <- sig_genes[sig_genes %in% GKN2]

#Visualizing KEGG Results- Pathview
#Pathview is a tool set for pathway based data integration and visualization. It maps and renders a wide variety of biological data on relevant pathway graphs.

#Let's take a look at our top KEGG pathway in Pathview
geneData <- as.numeric(Seq_sig$logFC)
names(geneData) <- Seq_sig$ensembl_gene_id

options(bitmapType='cairo', warn=-1)
pdf(file="hsa04080.RNA.pdf")
pathways <- pathview(gene.data = geneData, pathway.id = "04080", species = "hsa", gene.idtype="ENSEMBL", out.suffix = "RNA")

#And our second most enriched pathway
options(bitmapType='cairo', warn=-1)
pdf(file="hsa05033.RNA.pdf")
pathways <- pathview(gene.data = geneData, pathway.id = "05033", species = "hsa", gene.idtype="ENSEMBL", out.suffix = "RNA")
#END
